lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -612,26 +840,34 @@ void CreateSiFiveMetadataPass::runOnOperation() {
          "cannot handle more than one CircuitOp in a mlir::ModuleOp");
 
   auto *body = circuitOp.getBodyBlock();
-
   // Find the device under test and create a set of all modules underneath it.
   auto it = llvm::find_if(*body, [&](Operation &op) -> bool {
     return AnnotationSet(&op).hasAnnotation(dutAnnoClass);
   });
+  auto &instanceGraph = getAnalysis<InstanceGraph>();
   if (it != body->end()) {
     dutMod = dyn_cast<FModuleOp>(*it);
-    auto &instanceGraph = getAnalysis<InstanceGraph>();
     auto *node = instanceGraph.lookup(cast<igraph::ModuleOpInterface>(*it));
     llvm::for_each(llvm::depth_first(node),
                    [&](igraph::InstanceGraphNode *node) {
                      dutModuleSet.insert(node->getModule());
                    });
   }
-  ObjectModelIR omir(moduleOp);
+  ObjectModelIR omir(circuitOp, dutMod, instanceGraph, moduleNamespaces);
 
   if (failed(emitRetimeModulesMetadata(omir)) ||
       failed(emitSitestBlackboxMetadata(omir)) ||
       failed(emitMemoryMetadata(omir)))
     return signalPassFailure();
+  auto *node = instanceGraph.getTopLevelNode();
+  if (FModuleOp topMod = dyn_cast<FModuleOp>(*node->getModule()))
+    if (auto objectOp = omir.instantiateSifiveMetadata(topMod)) {
+      SmallVector<std::pair<unsigned, PortInfo>> ports = {
+          {topMod.getNumPorts(),
+           PortInfo(StringAttr::get(objectOp->getContext(), "metadataObj"),
+                    objectOp.getType(), Direction::Out)}};
+      topMod.insertPorts(ports);
+    }
 
   // This pass does not modify the hierarchy.
   markAnalysesPreserved<InstanceGraph>();
```

Comment:
Since FIRRTL classes are FModuleLike, I think the new classes and objects added by this path do affect the instance graph. So we should either update the InstanceGraph as we create classes and objects, or not mark that analysis preserved.

Reply:
Thanks for spotting this, removed the `markAnalysesPreserved`.

Reply:
Sounds good, I think we can move forward with this now that it's safe. If re-computing the instance graph after this pass becomes painful we know what we have to do to keep it up to date.